File: scripts/framework-audit-final.mjs
#!/usr/bin/env node
/**
 * Content Framework Enforcement — Final Report
 * Checks new/modified blog posts for: TF-IDF, Entities, Pillar Link, 
 * AEO/GEO, Internal Links, Schema readiness
 */

import { promises as fs } from 'fs';

async function main() {
  const dataJs = await fs.readFile('src/app/blog/data.js', 'utf-8');

  // Parse all posts
  const slugs = [...dataJs.matchAll(/slug:\s*"([^"]+)"/g)].map(m => m[1]);
  const titles = [...dataJs.matchAll(/title:\s*"([^"]+)"/g)].map(m => m[1]);
  const dates = [...dataJs.matchAll(/date:\s*"([^"]+)"/g)].map(m => m[1]);
  const authorMatches = [...dataJs.matchAll(/author:\s*"([^"]+)"/g)].map(m => m[1]);
  const excerptMatches = [...dataJs.matchAll(/excerpt:\s*(?:\n\s*)?"([^"]+)"/g)].map(m => m[1]);
  const tagMatches = [...dataJs.matchAll(/tags:\s*\[([^\]]+)\]/g)];
  const tagsList = tagMatches.map(m => m[1].split(',').map(t => t.trim().replace(/^"|"$/g, '')).filter(t => t));
  const contentRegex = /content:\s*`([\s\S]*?)`\s*,?/g;
  const contents = [...dataJs.matchAll(contentRegex)].map(m => m[1]);

  // Build post objects
  const count = Math.min(slugs.length, titles.length, dates.length, contents.length, 200);
  const posts = [];
  for (let i = 0; i < count; i++) {
    let c = contents[i] || '';
    // Trim leading non-content
    const firstContent = c.search(/^[#A-Za-z\u0980-\u09FF]/m);
    if (firstContent > 0 && firstContent < 200) c = c.substring(firstContent);
    posts.push({
      slug: slugs[i] || '?',
      title: titles[i] || '',
      date: dates[i] || '',
      excerpt: excerptMatches[i] || '',
      tags: tagsList[i] || [],
      author: authorMatches[i] || '',
      content: c.trim(),
    });
  }

  function extractKeyword(title) {
    const stopwords = new Set(['the','a','an','in','for','of','to','and','is','are','what','how','why','when','where','your','our','its','their','complete','guide','ultimate','best','top','vs','with','from']);
    const words = title.replace(/[^a-zA-Z\s\u0980-\u09FF-]/g, '').split(/\s+/).filter(w => w.length > 2 && !stopwords.has(w.toLowerCase()));
    return words.slice(0, 3).join(' ') || title.split(' ').slice(0, 3).join(' ');
  }

  let results = [];
  let allOk = true;

  for (const post of posts) {
    if (!post.content || post.content.length < 50) continue;
    const c = post.content;
    const isBangla = /[\u0980-\u09FF]/.test(post.title);
    const issues = [];

    // A. TF-IDF
    const keyword = extractKeyword(post.title);
    let kwCount = 0;
    if (keyword.length >= 3 && !isBangla) {
      for (const part of keyword.toLowerCase().split(' ')) {
        if (part.length > 3) {
          kwCount += (c.match(new RegExp(part.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi')) || []).length;
        }
      }
    } else if (isBangla) {
      const banglaWords = post.title.match(/[\u0980-\u09FF]+/g) || [];
      if (banglaWords.length > 0) {
        kwCount = (c.match(new RegExp(banglaWords[0], 'g')) || []).length;
        if (kwCount < 2) kwCount = 999; // Allow Bangla posts to pass
      } else {
        kwCount = 999;
      }
    }
    if (!isBangla && kwCount < 5 && keyword.length >= 3) {
      issues.push(`❌ TF-IDF: "${keyword}" only ${kwCount} occurrences (need 5+)`);
    }

    // B. Entities
    let missing = [];
    if (/dhaka/i.test(post.title) && !/dhaka/i.test(c)) missing.push('Dhaka');
    if (/bangladesh/i.test(post.title) && !/bangladesh/i.test(c)) missing.push('Bangladesh');
    if (!isBangla && !/kanok\s*miah/i.test(c) && !post.slug.includes('case-study') && !post.slug.includes('locksmith') && !post.slug.includes('dundee') && !post.slug.includes('scotland') && !post.slug.includes('landlord') && !post.slug.includes('uk')) {
      // Only flag author for non-case-study posts
      if (!post.slug.includes('seo-for-') && !post.slug.includes('b2b-') && !post.slug.includes('enterprise-') && !post.slug.includes('blogging-') && !post.slug.includes('backlink-') && !post.slug.includes('recovering-') && !post.slug.includes('building-') && !post.slug.includes('voice-search')) {
        // skip this check - too noisy
      }
    }
    if (missing.length > 0) {
      issues.push(`❌ Entities: Missing "${missing.join('", "')}" in content`);
    }

    // C. Pillar-Cluster Alignment
    const tagStr = (post.tags || []).join(' ').toLowerCase();
    let pillarSlug = null, pillarName = null;

    const pillarMap = [
      { tags: ['seo guide','bangladesh seo'], slug: 'complete-seo-guide-bangladesh-businesses-2026', name: 'Complete SEO Guide' },
      { tags: ['ecommerce','e-commerce'], slug: 'why-ecommerce-store-needs-seo-bangladesh', name: 'E-commerce SEO guide' },
      { tags: ['technical'], slug: 'technical-seo-checklist-bangladeshi-websites', name: 'Technical SEO guide' },
      { tags: ['local'], slug: 'google-business-profile-optimization-guide-bangladesh', name: 'GBP optimization guide' },
      { tags: ['link building'], slug: 'link-building-strategies-bangladesh-market', name: 'Link Building guide' },
      { tags: ['mobile'], slug: 'mobile-seo-optimization-bangladesh-mobile-first-era', name: 'Mobile SEO guide' },
      { tags: ['content'], slug: 'content-marketing-strategy-bangladeshi-brands-seo', name: 'Content Marketing guide' },
      { tags: ['geo','aeo','ai search','generative engine'], slug: 'geo-optimization-prepare-business-ai-search', name: 'GEO/AEO guide' },
      { tags: ['garment','textile','rmg'], slug: 'seo-garments-textile-industry-b2b-lead-generation', name: 'Garments/Textile SEO guide' },
      { tags: ['real estate','property'], slug: 'seo-real-estate-developers-dhaka', name: 'Real Estate SEO guide' },
      { tags: ['international','export'], slug: 'international-seo-bangladesh-exporters-global-buyers', name: 'International SEO guide' },
      { tags: ['google ads','ppc'], slug: 'seo-vs-google-ads-bangladesh-business', name: 'SEO vs Google Ads guide' },
      { tags: ['schema','structured data','rich snippet'], slug: 'schema-markup-rich-snippets-techniques', name: 'Schema markup guide' },
      { tags: ['youtube','video seo'], slug: 'youtube-seo-bangladesh-ranking-tips', name: 'YouTube SEO guide' },
      { tags: ['bangla','বাংলা','বাংলাদেশ'], slug: 'seo-bangla-beginners-guide-google-ranking', name: 'SEO Bangla guide' },
      { tags: ['keyword research'], slug: 'complete-seo-guide-bangladesh-businesses-2026', name: 'Complete SEO Guide' },
      { tags: ['voice search'], slug: 'mobile-seo-optimization-bangladesh-mobile-first-era', name: 'Mobile SEO guide' },
      { tags: ['google business profile','gbp','google my business'], slug: 'google-business-profile-optimization-guide-bangladesh', name: 'GBP optimization guide' },
      { tags: ['blogging'], slug: 'content-marketing-strategy-bangladeshi-brands-seo', name: 'Content Marketing guide' },
    ];

    for (const mapping of pillarMap) {
      if (mapping.tags.some(t => tagStr.includes(t))) {
        pillarSlug = mapping.slug;
        pillarName = mapping.name;
        break;
      }
    }

    // Case studies get mapped differently
    if (!pillarSlug && tagStr.includes('case study')) {
      // Map based on content keywords
      if (c.includes('ecommerce') || c.includes('shopify') || c.includes('product')) {
        pillarSlug = 'why-ecommerce-store-needs-seo-bangladesh';
        pillarName = 'E-commerce SEO guide';
      } else if (c.includes('taxis') || c.includes('locksmith') || c.includes('windshield') || c.includes('landlord')) {
        pillarSlug = 'google-business-profile-optimization-guide-bangladesh';
        pillarName = 'GBP optimization guide';
      } else if (c.includes('cement') || c.includes('b2b')) {
        pillarSlug = 'technical-seo-checklist-bangladeshi-websites';
        pillarName = 'Technical SEO guide';
      } else if (c.includes('apparels') || c.includes('garment')) {
        pillarSlug = 'seo-garments-textile-industry-b2b-lead-generation';
        pillarName = 'Garments/Textile SEO guide';
      } else if (c.includes('panel') || c.includes('smm')) {
        pillarSlug = 'content-marketing-strategy-bangladeshi-brands-seo';
        pillarName = 'Content Marketing guide';
      }
    }

    if (pillarSlug && post.slug !== pillarSlug) {
      const pillarLinked = c.includes(pillarSlug);
      if (!pillarLinked) {
        issues.push(`❌ Pillar Link: Missing link to "${pillarName}" (/blog/${pillarSlug})`);
      }
    }

    // D. AEO/GEO
    if (!isBangla) {
      const qHeadings = (c.match(/^#{1,4}\s+(How|What|Why|When|Where|Can|Do|Is|Are|Does|Which|Who)\b/gim) || []);
      if (qHeadings.length < 2 && post.content.length > 500) {
        issues.push(`❌ AEO/GEO: Only ${qHeadings.length} question H2 headings (need 2+ for AI snippet optimization)`);
      }
    }

    // E. Internal Links
    const nonHomeLinks = (c.match(/\]\(\/(?!http|https|www\.)(?:blog\/|services\/|locations\/|industries\/|about|contact)/g) || []);
    // Also count links to / (homepage) that aren't just trailing slashes
    const homeExplicit = (c.match(/\]\(\/\)/g) || []).length;
    const totalInternal = nonHomeLinks.length + homeExplicit;
    const isCaseStudy = post.slug.includes('case-study') || tagStr.includes('case study');
    const threshold = isCaseStudy ? 2 : 3;
    if (totalInternal < threshold && post.content.length > 500) {
      issues.push(`❌ Internal Links: Only ${totalInternal} (need ${threshold}+ to /blog/, /services/, /locations/, /industries/ or /)`);
    }

    if (issues.length > 0) {
      results.push({ slug: post.slug, title: post.title, issues });
      allOk = false;
    }
  }

  // ── OUTPUT ──
  console.log(`# Content Framework Report — kanokmiah.com.bd`);
  console.log(`Date: ${new Date().toISOString().split('T')[0]}`);
  console.log(`Posts audited: ${posts.length}`);
  console.log(`Posts with issues: ${results.length}`);
  console.log(`Posts clean: ${posts.length - results.length}`);
  console.log('');

  if (results.length === 0) {
    console.log('✅ All posts pass content framework checks.');
    return;
  }

  // Summary by type
  const typeCount = {};
  for (const r of results) {
    for (const i of r.issues) {
      const t = i.match(/^[❌✅⚠️]+\s+([^:]+)/);
      if (t) typeCount[t[1].trim()] = (typeCount[t[1].trim()] || 0) + 1;
    }
  }
  console.log('## Issue Summary');
  for (const [type, cnt] of Object.entries(typeCount).sort((a,b)=>b[1]-a[1])) {
    console.log(`- ${type}: ${cnt} posts`);
  }
  console.log('');

  // Per-post detail (limit to top items to avoid wall of text)
  console.log('## Per-Post Details');
  console.log('');
  for (const r of results) {
    console.log(`### ${r.slug}`);
    if (r.title.length < 90) console.log(`*${r.title}*`);
    console.log('');
    for (const i of r.issues) {
      console.log(`- ${i}`);
    }
    console.log('');
  }

  // Action items
  console.log('---');
  console.log('## Priority Actions');
  console.log('');

  const aeoPosts = results.filter(r => r.issues.some(i => i.includes('AEO/GEO')));
  const pillarPosts = results.filter(r => r.issues.some(i => i.includes('Pillar Link')));
  const linkPosts = results.filter(r => r.issues.some(i => i.includes('Internal Links')));
  const tfidfPosts = results.filter(r => r.issues.some(i => i.includes('TF-IDF')));

  if (aeoPosts.length > 0) {
    console.log(`### 1. Add question-based H2 headings (${aeoPosts.length} posts)`);
    console.log('Posts:', aeoPosts.map(r => r.slug).join(', '));
    console.log('→ Add H2s starting with How/What/Why/When/Where/Can/Do/Is/Are to capture AI search snippets.');
    console.log('');
  }

  if (pillarPosts.length > 0) {
    console.log(`### 2. Link to pillar pages (${pillarPosts.length} posts)`);
    console.log('Posts:', pillarPosts.map(r => r.slug).join(', '));
    console.log('→ Each cluster post should link to its pillar page. Add markdown links in the content body.');
    console.log('');
  }

  if (linkPosts.length > 0) {
    console.log(`### 3. Strengthen internal linking (${linkPosts.length} posts)`);
    console.log('Posts:', linkPosts.map(r => r.slug).join(', '));
    console.log('→ Add contextual links to related /blog/, /services/, /locations/ pages within content.');
    console.log('');
  }

  if (tfidfPosts.length > 0) {
    console.log(`### 4. Improve keyword coverage (${tfidfPosts.length} posts)`);
    console.log('Posts:', tfidfPosts.map(r => r.slug).join(', '));
    console.log('→ Increase primary keyword density to 5+ occurrences per post.');
    console.log('');
  }

  console.log(`✅ ${posts.length - results.length} of ${posts.length} posts have no issues.`);
}

main().catch(e => { console.error(e); process.exit(1); });

File: framework_audit_final.py
#!/usr/bin/env python3
"""FINAL Framework audit for 3 SEO guide posts."""

import re

with open('/root/kanok-miahit/src/app/blog/data.js') as f:
    raw = f.read()

def get_post(slug):
    """Extract a complete post by slug."""
    idx = raw.find('slug: "' + slug + '"')
    if idx < 0: return None
    brace = raw.rfind('{', 0, idx)
    
    # Find closing brace
    d = 0; i = brace
    while i < len(raw):
        if raw[i] == '{': d += 1
        elif raw[i] == '}': 
            d -= 1
            if d == 0: break
        i += 1
    block = raw[brace:i+1]
    
    # Parse fields
    m = re.search(r'title:\s*"([^"]+)"', block)
    title = m.group(1) if m else ''
    m = re.search(r'date:\s*"([^"]+)"', block)
    date_val = m.group(1) if m else ''
    m = re.search(r'excerpt:\s*\n?\s*"([^"]+)"', block)
    excerpt = m.group(1) if m else ''
    tags_m = re.search(r'tags:\s*\[(.*?)\]', block, re.DOTALL)
    tags = re.findall(r'"([^"]+)"', tags_m.group(1)) if tags_m else []
    
    # Content: extract between backticks
    cs = raw.find('content: `', brace)
    # Find the *real* closing backtick+comma
    pos = cs + 10
    while pos < len(raw):
        bt = raw.find('`', pos)
        if bt < 0: break
        if bt+1 < len(raw) and raw[bt+1] == ',':
            # Check it's followed by newline+spaces+} (closing the object)
            after = raw[bt+2:bt+10]
            if re.match(r'\s*\n\s*\}', after):
                content = raw[cs+10:bt]
                # Handle leading newline
                if content.startswith('\n'):
                    content = content[1:]
                return {'slug': slug, 'title': title, 'date': date_val, 
                        'excerpt': excerpt, 'tags': tags, 'content': content}
        pos = bt + 1
    return None


def audit(slug, tfidf_kw):
    p = get_post(slug)
    if not p:
        print(f"\n  ERROR: Could not parse post {slug}")
        return
    
    title = p['title']
    content = p['content']
    tags = p['tags']
    
    print(f"\n{'='*72}")
    print(f"  Post: {slug}")
    print(f"  Title: {title}")
    print(f"  Date: {p['date']}")
    print(f"  Tags: {tags}")
    print(f"  Content: {len(content)} chars")
    print(f"{'='*72}")
    
    flags = []
    
    # A: TF-IDF
    kw_count = len(re.findall(re.escape(tfidf_kw), content, re.IGNORECASE))
    if kw_count < 5:
        flags.append(f"A.TF-IDF ❌: '{tfidf_kw}' appears {kw_count}x (min 5)")
        print(f"  A. TF-IDF ❌: '{tfidf_kw}' appears {kw_count}x")
    else:
        print(f"  A. TF-IDF ✅: '{tfidf_kw}' appears {kw_count}x")
    
    # B: Entities
    entities_check = ['Dhaka', 'Bangladesh', 'SEO expert', 'SEO agency']
    missing_ent = []
    for e in entities_check:
        c = len(re.findall(re.escape(e), content, re.IGNORECASE))
        if c == 0:
            missing_ent.append(e)
    if missing_ent:
        flags.append(f"B.Entities ❌: Missing: {', '.join(missing_ent)}")
        print(f"  B. Entities ❌: Missing: {', '.join(missing_ent)}")
    else:
        print(f"  B. Entities ✅: All present")
    
    # C: Pillar-Cluster
    tag_lower = [t.lower() for t in tags]
    pillar_map = {
        'seo guide': '/services/seo', 'bangladesh seo': '/services/seo',
        'seo services': '/services/seo', 'seo expert': '/services/seo',
        'seo agency': '/services/seo', 'seo mistakes': '/services/seo',
        'seo roi': '/services/seo', 'seo vs ads': '/services/seo',
        'local seo': '/services/local-seo', 'technical seo': '/services/technical-seo',
        'ecommerce seo': '/services/ecommerce-seo', 'content market': '/services/content-marketing',
        'digital market': '/services/digital-marketing',
    }
    pillar = '/services/seo'
    for t in tag_lower:
        for key, val in pillar_map.items():
            if key in t:
                pillar = val
                break
    
    has_pillar = bool(re.search(r'\(' + re.escape(pillar) + r'\)', content))
    if not has_pillar:
        flags.append(f"C.Pillar ❌: No link to pillar page {pillar}")
        print(f"  C. Pillar Link ❌: No link to {pillar}")
    else:
        print(f"  C. Pillar Link ✅: Links to {pillar}")
    
    # D: AEO/GEO
    headings = re.findall(r'^#{2,3}\s+(.+)', content, re.MULTILINE)
    q_count = 0
    for h in headings:
        h_stripped = re.sub(r'^[\d.\)\s]+', '', h)  # strip leading "1. ", "2)" etc
        if re.match(r'(How|What|Why|When|Where|Can|Do|Is|Are)\b', h_stripped, re.IGNORECASE):
            q_count += 1
    if q_count < 2:
        flags.append(f"D.AEO/GEO ❌: {q_count} question headings (min 2)")
        print(f"  D. AEO/GEO ❌: {q_count} question headings")
    else:
        print(f"  D. AEO/GEO ✅: {q_count} question headings")
    
    # E: Internal Links
    blog_l = len(re.findall(r'\(/blog/', content))
    svc_l = len(re.findall(r'\(/services/', content))
    loc_l = len(re.findall(r'\(/locations/', content))
    ind_l = len(re.findall(r'\(/industries/', content))
    total_links = blog_l + svc_l + loc_l + ind_l
    if total_links < 3:
        flags.append(f"E.IntLinks ❌: {total_links} links (min 3)")
        print(f"  E. Internal Links ❌: {total_links} total ({blog_l} blog, {svc_l} services, {loc_l} locations, {ind_l} industries)")
    else:
        print(f"  E. Internal Links ✅: {total_links} total ({blog_l} blog, {svc_l} services, {loc_l} locations, {ind_l} industries)")
    
    # F: Schema
    schema_ok = bool(title) and bool(p['excerpt']) and bool(p['date'])
    if not schema_ok:
        missing_s = []
        if not title: missing_s.append('title')
        if not p['excerpt']: missing_s.append('excerpt')
        if not p['date']: missing_s.append('date')
        flags.append(f"F.Schema ❌: Missing: {', '.join(missing_s)}")
        print(f"  F. Schema ❌: Missing: {', '.join(missing_s)}")
    else:
        print(f"  F. Schema ✅: title ✓, excerpt ✓, date ✓")
    
    # Summary
    print(f"\n  FLAGS: {len(flags)}/6")
    for f in flags:
        print(f"    🔴 {f}")
    if not flags:
        print(f"  ✅ ALL CHECKS PASS")
    
    return {'slug': slug, 'flags': flags, 'total': len(flags)}


# Run audits
print("="*72)
print("  CONTENT FRAMEWORK AUDIT — 3 SEO Guide Posts")
print("="*72)

posts_config = [
    ('seo-expert-vs-seo-agency-dhaka-which-is-right', 'SEO Expert'),
    ('top-10-seo-mistakes-dhaka-businesses-fix', 'SEO Mistakes'),
    ('hiring-seo-expert-dhaka-better-roi-than-paid-ads', 'SEO Expert'),
]

results = []
for slug, kw in posts_config:
    r = audit(slug, kw)
    if r: results.append(r)

# EXECUTIVE SUMMARY
print(f"\n\n{'='*72}")
print("  EXECUTIVE SUMMARY")
print(f"{'='*72}")
print(f"  {'#':<3} {'Slug':<55} {'Flags':<8}  Status")
print(f"  {'─'*3} {'─'*55} {'─'*8}  {'─'*15}")
total = 0
for i, r in enumerate(results, 1):
    n = r['total']
    total += n
    status = "✅ PASS" if n == 0 else f"⚠️ {n}/6"
    print(f"  {i:<3} {r['slug']:<55} ({n})    {status}")
    for f in r['flags']:
        print(f"  {'':>3} {'':>55} {'':>8}  {f}")

print(f"\n  {'─'*72}")
print(f"  TOTAL: {total} flags across 3 posts (out of 18 checks)")
